### PR TITLE
Feat: Change current track's favorited state

### PR DESCRIPTION
**Pull request by ducktordanny**
With the included changes we'll be able to mark a track as favorited or undo this via the following examples:

- Favorite: `require('apple-music').set_current_track_favorited(true)`
- Undo favorite: `require('apple-music').set_current_track_favorited(false)`

This command will also handle if the OS version is below 14 (Sonoma) where the `favorited` property was called `loved`. This is done via the `osascript` that @p5quared provided in the related [issue](https://github.com/p5quared/apple-music.nvim/issues/10)'s comment.

If the command succeeds it will prompt the change and the name plus the author of the track it was applied for.

With these changes, we're now also able to get the current track with `require('apple-music').get_current_track()` and it will return the name and author of the track in the following format `<name> - <author>`.

### CODE REVIEW

File: lua/apple-music/init.lua [hunk 1]
```diff
@@ -180,6 +186,34 @@ M.disable_shuffle = function()
 	end
 end
 
+---Change the favorited state of the current music.
+---It also handles Mac OS versions below 14 (Sonoma) where it used to be called "loved".
+---@param state boolean: The state of favorited to be set for current track.
+---@usage require('apple-music').set_current_track_favorited(true)
+---@usage require('apple-music').set_current_track_favorited(false)
+M.set_current_track_favorited = function(state)
```
Comment (p5quared) [inline lua/apple-music/init.lua:194]: I think this functionality would be better if we split it into three:

1. `favorite_current()`
2. `unfavorite_current()`
3. `toggle_favorite_track()` (If we can get "is_favorited")

I think most people would use a keybinding and not want to worry about parameters. 

Don't have to worry about `toggle_favorite_track`, that can be a separate PR.

**Review by p5quared — COMMENTED** [review]
I love this feature!

I left some notes. Here's what I'd like to see to merge:
1. `get_current_track` should just return the name, and leave it to users to decide what they want.
2. Split up `set_current_track_favorited` to `favorite_current_track` and `unfavorite_current_track`.

Optional:
What do you think about naming the split functions just `favorite()` and `unfavorite()`?
I don't think you can favorite anything besides tracks, and I doubt anyone will be trying to favorite/unfavorite anything besides the currently playing track.
In the future we could add `favorite_track_by_name(name)` or something if it is desired.

I left some other small suggestions like `get_current_artist`, I actually think I'd prefer those in separate PRs.
I'm active pretty much every day and am a fan of many tiny easy PRs over smushing everything together.

File: lua/apple-music/init.lua (hunk 1 above)
Comment (ducktordanny) [reply lua/apple-music/init.lua:194]: Yeah, probably it'd be simpler to map this way. The favorited property can be fetched from the current track so the toggle should be feasible. 👍🏼

**Comment by ducktordanny** [comment]
As I've just checked you can also favorite playlists and albums as well. But I tried this in the app (iPhone) and it seems those don't show up in the built-in favorite songs playlist  (I assume it only fetches the favorite tracks in the background), which is weird because then we can't see the favorited albums and playlists in a summary (at least in the app). Anyway, I think this should be included in another PR later on.

File: lua/apple-music/init.lua [hunk 2]
```diff
@@ -44,6 +44,12 @@ local execute = function(cmd)
 	return pcall(exe, cmd)
 end
 
+local grab_os_version = function()
+	local command = [[ osascript -e 'set osver to system version of (system info)' ]]
+	local _, result = execute(command)
+	return tonumber(result)
```
Comment (ducktordanny) [inline lua/apple-music/init.lua:50]: A version in x.y.z format can break this.

Comment (p5quared) [reply lua/apple-music/init.lua:50]: Good catch! 
We only need the major version.

File: lua/apple-music/init.lua [hunk 3]
```diff
@@ -328,6 +362,20 @@ M.select_album_telescope = function()
 	}):find()
 end
 
+---Get the name and artist of the current track (in the following format: <name> - <author>).
```
[the remark this answers is not in the document]
Comment (p5quared) [reply lua/apple-music/init.lua:365]: accidentally started 2 reviews yesterday.

[the remark this answers is not in the document]
Comment (p5quared) [reply lua/apple-music/init.lua:365]: This was from then

File: lua/apple-music/init.lua (hunk 2 above)
Comment (ducktordanny) [reply lua/apple-music/init.lua]: Done.

File: lua/apple-music/init.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (ducktordanny) [reply lua/apple-music/init.lua]: > I'd prefer if you'd limit this to just get_current_track() -> trackname.

Do you mean it should be `get_current_trackname()`?

File: lua/apple-music/init.lua (hunk 1 above)
Comment (ducktordanny) [reply lua/apple-music/init.lua]: I still kind of left this function there, but as a local one so we have less duplication and it can be also reused later on. But now we have two functions for this.

File: lua/apple-music/init.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (p5quared) [reply lua/apple-music/init.lua]: Sorry this was a duplicate I lost and then accidentally posted.

I was referring to the return just being `trackname`

**Comment by p5quared** [comment]
@ducktordanny When you're all set just re-request review from me.

File: lua/apple-music/init.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (ducktordanny) [reply lua/apple-music/init.lua]: Ohhhh, okay, got it 😃

File: lua/apple-music/init.lua [hunk 4]
```diff
@@ -180,6 +209,28 @@ M.disable_shuffle = function()
 	end
 end
 
+---Favorite current track.
+---NOTE: Below Sonoma it'll be loved.
```
Comment (p5quared) [inline lua/apple-music/init.lua:213]:
```suggestion
---Favorite current (playing) track.
---NOTE: Below macOS 14 (Sonoma), it'll be 'loved'.
```

File: lua/apple-music/init.lua [hunk 5]
```diff
@@ -180,6 +209,28 @@ M.disable_shuffle = function()
...
+M.favorite_current_track = function()
+	local current_track = M.get_current_trackname()
+	if not current_track then
+		return
+	end
+	set_track_favorited_state(current_track, true)
+end
+
+---Unfavorite current track.
+---NOTE: Below Sonoma it'll be unloved.
```
Comment (p5quared) [inline lua/apple-music/init.lua:224]:
```suggestion
---Unfavorite current (playing) track.
---NOTE: Below macOS 14 (Sonoma), it'll be un-'loved'.
```

File: lua/apple-music/init.lua [hunk 6]
```diff
@@ -44,6 +44,35 @@ local execute = function(cmd)
 	return pcall(exe, cmd)
 end
 
+local grab_major_os_version = function()
+	local command = [[ osascript -e 'set osver to system version of (system info)' ]]
+	local _, result = execute(command)
+	return tonumber(result:match("%d+"))
+end
+
+---Change the favorited state of a track.
```
Comment (p5quared) [inline lua/apple-music/init.lua:53]:
```suggestion
---Set the favorited/loved state of a track.
```

File: lua/apple-music/init.lua [hunk 7]
```diff
@@ -328,6 +379,18 @@ M.select_album_telescope = function()
 	}):find()
 end
 
+---Get the name of the current track.
+---@usage require('apple-music').get_current_trackname()
+M.get_current_trackname = function()
```
Comment (p5quared) [inline lua/apple-music/init.lua:384]: I'd rather this be a local function rather than exported with the module. My reason is because I'd like to add a function to get `current_track_info()` and this will be unnecessary when that is around.

It will also help us if we split up the module a bit more like you suggested in discussion.

**Review by p5quared — APPROVED** [review]
Looks great! :heart:

You are welcome to merge as is and I can do the minor refactor (moving the `get_current_trackname()` outside of `M`) if you want. Otherwise, if you do that refactor I'll also approve. 

I also reworded some comments. Feel free to disagree if you feel strongly about them.

File: lua/apple-music/init.lua (hunk 7 above)
Comment (ducktordanny) [reply lua/apple-music/init.lua:384]: Okay, that sounds fine.

**Comment by ducktordanny** [comment]
Made the suggested changes I think those make it better. 👌

**Comment by p5quared** [comment]
Awesome! 

I'm merging this as is. I haven't really tested it much so I'll do that tonight and will put it in the next release probably with something for #17 as well.
